src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs

```diff
@@ -2293,8 +2297,8 @@ internal static bool IsAssemblyFriendOfSerialization(Assembly assembly)
             {
                 string internalsVisibleAttributeAssemblyName = internalsVisibleAttribute.AssemblyName;
 
-                if (Regex.IsMatch(internalsVisibleAttributeAssemblyName, Globals.SimpleSRSInternalsVisiblePattern) ||
-                    Regex.IsMatch(internalsVisibleAttributeAssemblyName, Globals.FullSRSInternalsVisiblePattern))
+                if (s_simpleSRSInternalsVisibleRegex.IsMatch(internalsVisibleAttributeAssemblyName) ||
+                    s_fullSRSInternalsVisibleRegex.IsMatch(internalsVisibleAttributeAssemblyName))
```

Comment:
Regex already has an internal cache it uses, right?  Should that cache be augmented to support caching the whole Regex object rather than just parts of it?

Reply:
Your suggestion is tempting as it would reduce cpu utilization and GC pressure for customers that aren't aware of these pitfalls. The downside would be that the constant heap allocation would increase. Especially when you are evaluating Regex expressions only in the early stages of your code execution it would stay in the cache forever...

That said I think there's definitely a great benefit in expanding the cache to entire Regex objects! Do we have some caching mechanisms (e.g. like in ArrayPool) that release some of the cached items over time?

Reply:
> Do we have some caching mechanisms (e.g. like in ArrayPool) that release some of the cached items over time?

Isn't the cached already used by Regex an MRU cache that releases the oldest items when adding new ones once a fixed limit is hit?  I was just suggesting it might be a simple change to modify the cache to also store a Regex object, and then rather than having IsMatch allocate a Regex which then checks the cache for its internals, IsMatch could just check the cache directly.  That said, I'm not intimately familiar with the internals, so that might not work.

Reply:
Yes it is a MRU cache. My concern was that if we cache the whole Regex object instead of just the code, the constant heap allocation increases noticeably. Maybe I'm over cautious and the increase wouldn't be measureable.

Your idea should work, the Regex class itself isn't that complex.

Reply:
> My concern was that if we cache the whole Regex object instead of just the code, the constant heap allocation increases noticeably. Maybe I'm over cautious and the increase wouldn't be measureable.

I'm not particularly concerned about that.  From a quick experiment, it looks like a basic phone number regex already caches over 8K of data, and it looks like the regex object itself is 104 bytes.

Reply:
I see. I will look into the feasibility. (but not in this PR)

Reply:
> but not in this PR

I'm wondering whether this PR is still needed in such a case.

Reply:
I think it's not a bad idea to separately store regex instances if they are needed multiple times (at least in the BCL). You don't know if it's still in the cache the next time you would want to evaluate an expression...

Reply:
the cost of it are two reference pointer, nothing greedy.

Reply:
> I think it's not a bad idea to separately store regex instances if they are needed multiple times (at least in the BCL)

This goes back to your question about overhead, and when/how often these are used.  If they're, for example, just used a few times at the beginning of a program and then never again, adding these static fields will keep maybe 15-20K of additional state alive for the duration of the program, even if they would subsequently have been evicted from the regex cache.

Do you know what the usage pattern of these is?

Reply:
I agree, we definitely need the answer of the usage pattern first. I wanted to save myself time and let @shmao answer that, but maybe I will just profile it by myself now.

Reply:
> the cost of it are two reference pointer, nothing greedy.

No, the cost isn't just the two pointers; rather, it's holding on to _all_ of the state associated with the Regex instances.

Reply:
I totally agree with what you are saying. Just for reference here are other places where the described pattern is used:

https://source.dot.net/#System.Data.Common/System/Data/Common/DbConnectionOptions.Common.cs,61
https://source.dot.net/#System.Private.Xml/System/Xml/Serialization/SourceInfo.cs,26
https://source.dot.net/#System.Private.Xml/System/Xml/XmlConvert.cs,304

Reply:
> Just for reference here are other places where Regex instances are cached statically

Yeah, I'm not debating the pattern: if the instances are going to be used repeatedly, or are part of APIs that are likely to be used repeatedly, by all means, cache.  If that's the case here, caching explicitly makes sense, too.  If it's not, then we shouldn't.

Reply:
The intent of posting the links wasn't to weaken your argumentation but to provide additional places to investigate possible memory optimizations :)